### PR TITLE
Dataset index and slicing

### LINKED ISSUE #871: Add "skip" feature to dataset kwargs and dataset generator to start at a specific index
This can be very useful for testing. For instance, sample 3313 in the UCF101 test set is the largest sample, so it would be helpful to test directly on that without processing through the rest.

### PR DESCRIPTION
**Pull request by davidslater**
Fixes #871 

This PR does a few related things:
1) Renames "split_type" to "split" to make it clear that it is verbatim what is given to TFDS load (same kwarg name)
https://www.tensorflow.org/datasets/api_docs/python/tfds/load

2) Modifies scenarios so that configs can override the train and eval splits. This enables TFDS slicing operations on them such as `"test[:10]"` or `"train_clean100+train_clean360+train_other500"`:
https://www.tensorflow.org/datasets/splits#examples

3) Adds additional parsing of splits to enable individual indexing for a specific data point, such as "test[355]", and indexing based off of a list, "test[[1, 4, 5, 7, 9]]". Under the hood, these are translated to single-element slices. Advanced slicing operations involving a `step` parameter, such as "test[10:20:2]" are not permitted, and raise a NotImplementedError.

### CODE REVIEW

**Comment by davidslater** [comment]
Biggest code changes are in `datasets.py` and its associated test file.

**Comment by lcadalzo** [comment]
What's the intended behavior when overlapping splits are provided? E.g. `"train[1:3] + train[2]"`. The current behavior is to repeat elements, so in this case `train[2]` is returned in two consecutive batches.

**Comment by lcadalzo** [comment]
Is it of interest to allow for specifying negative indices when (1) specifying a single index and/or (2) when providing a list of indices?

(1) E.g, if I only want to evaluate on the penultimate sample, specifying `"train[-2]"` will cause an error. `"train[[-2:-1]]"`, however, will yield the one sample I want.

(2) E.g., if I want to evaluate on the first and penultimate sample, `"train[[0, -2]]"` will cause an error since only positive indices are supported when passing a list. And `"train[0] + train[[-2:-1]]"` would work.

**Comment by davidslater** [comment]
> What's the intended behavior when overlapping splits are provided? E.g. `"train[1:3] + train[2]"`. The current behavior is to repeat elements, so in this case `train[2]` is returned in two consecutive batches.

The intended behavior is simply to match the underlying TFDS behavior, which allows duplicate and out-of-order splits. I don't expect people to make much use of it, but it is much simpler on our end than trying to determine uniqueness.

**Comment by davidslater** [comment]
> Is it of interest to allow for specifying negative indices when (1) specifying a single index and/or (2) when providing a list of indices?
> 
> (1) E.g, if I only want to evaluate on the penultimate sample, specifying `"train[-2]"` will cause an error. `"train[-2:-1]"`, however, will yield the one sample I want.
> 
> (2) E.g., if I want to evaluate on the first and penultimate sample, `"train[[0, -2]]"` will cause an error since only positive indices are supported when passing a list. And `"train[0] + train[-2:-1]"` would work.

I had not considered using negative indices. I didn't realize they would work. I can make an update to allow them if you think that makes sense.

**Comment by lcadalzo** [comment]
I think, since it's unlikely they would be commonly used, we can hold off and only add later if it's something that a user is specifically looking for.

**Review by lcadalzo — APPROVED** [review]
LGTM
